acceptance: de-flake TestDecommission

tbg opened the pull request:
The test frequently got stuck during `decommission --wait` for a
freshly killed node, due to the following:

- node 1 gets killed
- node 3 increments node 1's epoch
- everything from now on happens on node 2
- calls IncrementEpoch, which stalls for now
- sets node 1's decommissioning flag
- gossip fires and updates node1's liveness map entry
- IncrementEpoch continues
- it fails (since the decommissioning flag mismatches)
- blindly updates the liveness[1], effectively clobbering its in-memory state
- calls to Decommission for node1 now turn into no-ops on node2, but
  when the liveness is queried, it looks like node1 isn't decommissioning.

Interestingly, when I looked at the liveness output as seen by the other two
nodes, they always had the same problem. Perhaps something else is at work on
top of the above? I filed an issue to investigate & fix the problematic behavior
of IncrementEpoch:

https://github.com/cockroachdb/cockroach/issues/18219

For this PR, I am simply using the semaphore to avoid interleavings between
IncrementEpoch and Decommission, and was able to run the test 10x without it
hanging (usually would hang every other time before, maybe more often).

Fixes #17995.

[1]: https://github.com/cockroachdb/cockroach/blob/a9ef8342681baa571ab552d2dc93185ac8eccb49/pkg/storage/node_liveness.go#L566-L572

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18220)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18220#-:-KtIHEU6GvM4tMZL1Hcn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
acceptance: de-flake TestDecommission

The test frequently got stuck during `decommission --wait` for a
freshly killed node, due to the following:

- node 1 gets killed
- node 3 increments node 1's epoch
- everything from now on happens on node 2
- calls IncrementEpoch, which stalls for now
- sets node 1's decommissioning flag
- gossip fires and updates node1's liveness map entry
- IncrementEpoch continues
- it fails (since the decommissioning flag mismatches)
- blindly updates the liveness[1], effectively clobbering its in-memory state
- calls to Decommission for node1 now turn into no-ops on node2, but
  when the liveness is queried, it looks like node1 isn't decommissioning.

Interestingly, when I looked at the liveness output as seen by the other two
nodes, they always had the same problem. Perhaps something else is at work on
top of the above? I filed an issue to investigate & fix the problematic behavior
of IncrementEpoch:

https://github.com/cockroachdb/cockroach/issues/18219

For this PR, I am simply using the semaphore to avoid interleavings between
IncrementEpoch and Decommission, and was able to run the test 10x without it
hanging (usually would hang every other time before, maybe more often).

Fixes #17995.

[1]: https://github.com/cockroachdb/cockroach/blob/a9ef8342681baa571ab552d2dc93185ac8eccb49/pkg/storage/node_liveness.go#L566-L572

tschottdorf commented:
Unfortunately hit the same flake despite the fix. I can see from gossip that all nodes have received the right info:

```
$ for port in 59743  59751 59761; do ./cockroach debug gossip-values --insecure --port $port | grep -F liveness:1; done
"liveness:1": {NodeID:1 Epoch:1 Expiration:1504645273.001476803,0 Draining:false Decommissioning:true}
"liveness:1": {NodeID:1 Epoch:1 Expiration:1504645273.001476803,0 Draining:false Decommissioning:true}
"liveness:1": {NodeID:1 Epoch:1 Expiration:1504645273.001476803,0 Draining:false Decommissioning:true}
```

but yet, none of the nodes reflect this in `node status --decommission`:

```
$ for port in 59743  59751 59761; do ./cockroach node status 1 --decommission --insecure --port $port; done
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
| id |     address     |      build      |     updated_at      |     started_at      | is_live | gossiped_replicas | is_decommissioning | is_draining |
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
|  1 | 127.0.0.1:59734 | 20aa80063-dirty | 2017-09-05 17:01:03 | 2017-09-05 17:01:03 | false   |                 1 | false              | false       |
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
(1 row)
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
| id |     address     |      build      |     updated_at      |     started_at      | is_live | gossiped_replicas | is_decommissioning | is_draining |
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
|  1 | 127.0.0.1:59734 | 20aa80063-dirty | 2017-09-05 17:01:03 | 2017-09-05 17:01:03 | false   |                 1 | false              | false       |
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
(1 row)
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
| id |     address     |      build      |     updated_at      |     started_at      | is_live | gossiped_replicas | is_decommissioning | is_draining |
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
|  1 | 127.0.0.1:59734 | 20aa80063-dirty | 2017-09-05 17:01:03 | 2017-09-05 17:01:03 | false   |                 1 | false              | false       |
+----+-----------------+-----------------+---------------------+---------------------+---------+-------------------+--------------------+-------------+
(1 row)
```

Looking.

tbg committed:
storage: don't throw the gossip liveness update away

Another indication that we should do
https://github.com/cockroachdb/cockroach/issues/18219.

tschottdorf commented:
Ok, added a commit that hopefully addresses the flakiness. It also explains why I was seeing the odd behavior on all nodes -- we were simply throwing the gossip update away in the common case.

tbg committed:
acceptance: deflake decommissioning event log test

The new bits of the test introduced in
https://github.com/cockroachdb/cockroach/pull/18178 were flaky since RPCs may be
sent to the down nodes.

tschottdorf commented:
OK, addressed one more flake introduced by https://github.com/cockroachdb/cockroach/pull/18178, though I wonder if it's entirely expected that these RPC errors leak to the client.

PTAL!

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/acceptance/decommission_test.go, line 288 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18220#-KtMmPXfAIuk2YuiB4S8:-KtMmPXfAIuk2YuiB4S9:bhiz3hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03759982495f26d7845b912221d7be589292432/pkg/acceptance/decommission_test.go#L288)):*
> ```Go
> 			// send RPCs to the (relatively recently) down node:
> 			//
> 			// pq: rpc error: code = Unavailable desc = grpc: the connection is
> ```

No, we don't want to leak these errors to the client. For idempotent operations, we should retry internally; for non-idempotent operations we should wrap them in AmbiguousResultError. The DistSender does this; is this error coming from somewhere other than the DistSender?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18220#-:-KtMml_RilZ_l8XGsVaS:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/acceptance/decommission_test.go, line 288 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/18220#-KtMmPXfAIuk2YuiB4S8:-KtNjkCdb2h8xSiqquGC:bio9l0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03759982495f26d7845b912221d7be589292432/pkg/acceptance/decommission_test.go#L288)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

No, we don't want to leak these errors to the client. For idempotent operations, we should retry internally; for non-idempotent operations we should wrap them in AmbiguousResultError. The DistSender does this; is this error coming from somewhere other than the DistSender?
</blockquote></details>

I think it would come from DistSender, but it's hard to tell. We should annotate these errors better. Filed https://github.com/cockroachdb/cockroach/issues/18264.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18220)*
<!-- Sent from Reviewable.io -->
